core/src/main/java/org/apache/calcite/plan/SubstitutionVisitor.java

```diff
@@ -192,8 +193,18 @@ public SubstitutionVisitor(RelNode target_, RelNode query_,
     this.simplify =
         new RexSimplify(cluster.getRexBuilder(), predicates, executor);
     this.rules = rules;
-    this.query = Holder.of(MutableRels.toMutable(query_));
+    MutableRel mutableQuery = MutableRels.toMutable(query_);
     this.target = MutableRels.toMutable(target_);
+    if (mutableQuery instanceof MutableCalc) {
+      this.query = Holder.of(mutableQuery);
+    } else {
```

Comment:
Here, we always try to compensate Calc in the original RelNode. Why can't we do it through custom normalization, or there will be more patterns in the future materialized view recognition. There should be a unified way to implement it.
How do you think?